### ISSUE TITLE
AsyncStyledComponent: A right way to manage dynamic styles

### ISSUE THREAD

**Issue by theKashey** [styled-components/styled-components#452]
So this is not about some bug, and even not about feature.
I come here from #134, look like I have a bit of knowledge how to fix a problem and speedup applying new styles.

So, where problem lies:
 On new component mount it is creating it`s own style, generating className and adds portion of text to current stylesheet.
 This lead to two problems: update of "global" stylesheet will reflow/redraw(in browser terms) all affected components(read - everything). Flush some useful caches and do some useless work. 

On each mount of each component.
Just because you need style to be available **instantly**. 

If you will await styles not right now, but in some point in future, everything can be a bit better.
So, do not apply styles asap, and dont delete them for a some time.
Apply new styles in bulk in new stylesheet, and then, after few seconds of idle, you can move new styles to global stylesheet.

The cons are simply - everything will be rendered in a strange way:
 - first everything will be rendered, and become invisible.
 - next styles will arrive to a browser, and whole tree will gain classNames. And regenerated.
 - then something changes style of component - they will not be changed instantly, but only after few ticks. Ie after flushing new styles to new stylesheet.

As result, it will change behavior how everything works, but may be it worth it.
May be not, cos it can break something.

Almost same logic are used in Yandex Maps API for handling dynamic styles. So it is a bit tested and proven solution. See a little demo: http://jsfiddle.net/gunwL8az/ 
PS: not a true example, cos each marker is SVG->PNG->DATAURI->STYLESHEET.

Sample code:
`
const createStyledComponent = (target: Target, rules: RuleSet, parent?: ReactClass<*>) => {

      generateAndInjectStyles(theme: any, props: any) {
        const executionContext = { ...props, theme }
        // some other implementation, flushes styles on nextTick, returns promise.
        --> return componentStyle._async_generateAndInjectStyles(executionContext)
      }

      componentWillMount() {
        // If there is a theme in the context, subscribe to the event emitter. This
        // is necessary due to pure components blocking context updates, this circumvents
        // that by updating when an event is emitted
        if (this.context[CHANNEL]) {
          const subscribe = this.context[CHANNEL]
          this.unsubscribe = subscribe(nextTheme => {
            // This will be called once immediately
            const { defaultProps } = this.constructor
            // Props should take precedence over ThemeProvider, which should take precedence over
            // defaultProps, but React automatically puts defaultProps on props.
            const isDefaultTheme = defaultProps && this.props.theme === defaultProps.theme
            const theme = this.props.theme && !isDefaultTheme ? this.props.theme : nextTheme
  
            // resolved via promise
            --> this.generateAndInjectStyles(theme, this.props).then( generatedClassName =>  
             this.setState({ theme, generatedClassName })
            );
          })
        } else {
          const theme = this.props.theme || {}
          this.generateAndInjectStyles(
            theme,
            this.props,
          --> ).then( generatedClassName =>  
             this.setState({ theme, generatedClassName })
            );
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe) {
         // are you sure this works?
          this.unsubscribe()
        }
      }

      render() {
        const { className, children, innerRef } = this.props
        let { generatedClassName } = this.state;

       if(!generatedClassName) {
          // we are not ready, but we cant stop rendering, cos in other case we will got "waves" of rendering. So:
         --> generatedClassName='_invisible_object_';
       }
     .....
        return createElement(target, propsForElement, children)
      }
    }
`

**Comment by jacobp100** [comment]
This would definitely be an interesting thing to benchmark if possible! We actually have a benchmarking branch—although at the moment it's mostly just testing the performance of parsing styles, not the application of styles.

**Comment by mxstbr** [comment]
Interesting—a WIP PR for this would be welcome so we can discuss further!

**Comment by theKashey** [comment]
Ok, will do my best.

**Comment by geelen** [comment]
Closing in favour of #553
